polymorphic.h

```diff
@@ -26,81 +26,110 @@ CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+      u_alloc_traits::deallocate(u_alloc, static_cast<U*>(src), 1);
+    };
+
+    clone = [](A& alloc, const T& src)->T* {
+      using u_allocator = typename std::allocator_traits<A>::template rebind_alloc<U>;
+      using u_alloc_traits = std::allocator_traits<u_allocator>;
+
+      u_allocator u_alloc(alloc);
+      auto* mem = u_alloc_traits::allocate(u_alloc, 1);
+      try {
+        u_alloc_traits::construct(u_alloc, mem, static_cast<const U&>(src));
+        return mem;
+      } catch (...) {
+        u_alloc_traits::deallocate(u_alloc, mem, 1);
+        throw;
+      }
+    };
+  }
+
+  void (*destroy)(A& alloc, T* src);
+  T* (*clone)(A& alloc, const T& src);
 };
 
-template <class T, class U, class A>
-class direct_control_block : public control_block<T, A> {
-  U u_;
+#if IMPL_OPTION == 1
 
- public:
-  template <class... Ts>
-  direct_control_block(Ts&&... ts) : u_(std::forward<Ts>(ts)...) {
-    control_block<T, A>::p_ = &u_;
-  }
+// Just initiate each (used) template variable instance. This is susceptible to static initialization order issues
+template<class T, class U, class A> inline control_block<T, A> s_control_block = control_block<T, A>(std::in_place_type<U>);
 
-  control_block<T, A>* clone(A& alloc) override {
-    using cb_allocator = typename std::allocator_traits<
-        A>::template rebind_alloc<direct_control_block<T, U, A>>;
-    cb_allocator cb_alloc(alloc);
-    using cb_alloc_traits = std::allocator_traits<cb_allocator>;
-    auto* mem = cb_alloc_traits::allocate(cb_alloc, 1);
-    try {
-      cb_alloc_traits::construct(cb_alloc, mem, u_);
-      return mem;
-    } catch (...) {
-      cb_alloc_traits::deallocate(cb_alloc, mem, 1);
-      throw;
-    }
-  }
+template<class T, class U, class A> control_block<T, A>* init_control_block()
+{
+  return &s_control_block<T, U, A>;
+}
 
-  void destroy(A& alloc) override {
-    using cb_allocator = typename std::allocator_traits<
-        A>::template rebind_alloc<direct_control_block<T, U, A>>;
-    cb_allocator cb_alloc(alloc);
-    using cb_alloc_traits = std::allocator_traits<cb_allocator>;
-    cb_alloc_traits::destroy(cb_alloc, this);
-    cb_alloc_traits::deallocate(cb_alloc, this, 1);
-  }
-};
+#elif IMPL_OPTION == 2
+
+// Put the control block inside the function to make sure it is initialized as early as needed. Properly thread-safe but the
+// overhead of checking the "initialized" flag for each call is a drawback.
+template<class T, class U, class A> control_block<T, A>* init_control_block()
+{
+  static control_block<T, A> instance(std::in_place_type<U>);
+  return &instance;
+}
+
+#elif IMPL_OPTION == 3
+
+template<class T, class U, class A> inline control_block<T, A> s_control_block;
+
+// Set up the control block each time to make sure it is set early enough without the function-static variable overhead.
+// Thread safety is so-so as multiple threads may write the same values to the function pointers at the same time, which could be
```

Comment:
Could we add some tests for this thread-safety concern? I'm not 100% sure I follow.

Reply:
I don't think it is a concern in reality, but probably officially UB. Each implementation would have to decide if it is ok to asynchronously write the same value to the same address in multiple cores. I have a hard time thinking of a hardware where this could fail, but in the "abstract machine" I'm pretty sure it is UB.

This design is however pretty old by now. The design I would go for (as an implementer) is now in cloning_ptr.h which allows up- and downcasting as per your original design that I found flawed. I don't understand why we didn't come up with this solution at the time. Maybe because we didn't want to accept a fat pointer _or_ an extra indirection. I now believe that you need to have either a fat pointer (like shared_ptr) or an indirection level (as in your polymorphic.h) but if you have it you can also allow casting as for shared_ptr.